### PR TITLE
New opcodes

### LINKED ISSUE #241: Redesign compound-type opcodes
```c#
public enum OpCode : byte
{
    PACK = 0xC0,
    UNPACK = 0xC1,
    NEWARRAY0 = 0xC2,
    NEWARRAY = 0xC3,
    NEWARRAY_T = 0xC4,
    NEWSTRUCT0 = 0xC5,
    NEWSTRUCT = 0xC6,
    //Reserved 0xC7
    NEWMAP = 0xC8,
    //Reserved 0xC9
    SIZE = 0xCA,
    HASKEY = 0xCB,
    KEYS = 0xCC,
    VALUES = 0xCD,
    PICKITEM = 0xCE,
    APPEND = 0xCF,
    SETITEM = 0xD0,
    REVERSEITEMS = 0xD1,
    REMOVE = 0xD2,
    CLEARITEMS = 0xD3,
}
```

### LINKED ISSUE #270: Type opcodes
```c#
public enum OpCode : byte
{
    ISNULL = 0xD8,
    ISTYPE = 0xD9,
    CONVERT = 0xDB,
}
```

### PR DESCRIPTION
**Pull request by erikzhang**
Closes #241, closes #270

### CODE REVIEW

**Comment by erikzhang** [comment]
Ready to review.

File: src/neo-vm/Types/ByteArray.cs [hunk 1]
```diff
@@ -5,29 +5,35 @@
...
 
         public ByteArray(ReadOnlyMemory<byte> value)
         {
-            this.memory = value;
+            this.Memory = value;
+        }
+
+        public override StackItem ConvertTo(StackItemType type)
+        {
+            if (type == StackItemType.Integer) return ToBigInteger();
```
Comment (shargon) [inline src/neo-vm/Types/ByteArray.cs]: ToBuffer?

File: src/neo-vm/Types/Buffer.cs [hunk 2]
```diff
@@ -21,6 +22,17 @@ public Buffer(ReadOnlySpan<byte> data)
...
 
+        public override StackItem ConvertTo(StackItemType type)
+        {
+            if (type == StackItemType.ByteArray)
+            {
+                byte[] clone = new byte[InnerBuffer.Length];
+                InnerBuffer.CopyTo(clone.AsSpan());
+                return clone;
+            }
+            return base.ConvertTo(type);
```
Comment (shargon) [inline src/neo-vm/Types/Buffer.cs]: ToInteger?

File: src/neo-vm/Types/ByteArray.cs (hunk 1 above)
Comment (shargon) [reply src/neo-vm/Types/ByteArray.cs]: Done

File: src/neo-vm/Types/Buffer.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo-vm/Types/Buffer.cs]: Done.

File: src/neo-vm/OpCode.cs [hunk 3]
```diff
@@ -741,11 +731,43 @@ public enum OpCode : byte
...
+        #endregion
+
+        #region Types
 
         /// <summary>
         /// Returns true if the input is null. Returns false otherwise.
         /// </summary>
-        ISNULL = 0xD0,
+        ISNULL = 0xD8,
+        [OperandSize(Size = 1)]
```
Comment (vncoelho) [inline src/neo-vm/OpCode.cs]: Perhaps add to other comments here, one for `ISTYPE` and one for `CONVERT`.

In summary, all changes looks good to me, readability and safety improved.

File: src/neo-vm/ExecutionEngine.cs [hunk 4]
```diff
@@ -852,8 +840,79 @@ private bool ExecuteInstruction()
...
+                        if (n < 0 || n > MaxStackSize) return false;
+                        StackItem item;
+                        if (instruction.OpCode == OpCode.NEWARRAY_T)
+                        {
+                            StackItemType type = (StackItemType)instruction.TokenU8;
+                            if (!Enum.IsDefined(typeof(StackItemType), type)) return false;
+                            item = instruction.TokenU8 switch
+                            {
+                                (byte)StackItemType.Integer => Integer.Zero,
+                                (byte)StackItemType.ByteArray => ByteArray.Empty,
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: Buffer?

File: src/neo-vm/ExecutionEngine.cs [hunk 5]
```diff
@@ -852,8 +840,79 @@ private bool ExecuteInstruction()
...
+                        }
+                        Push(new VMArray(ReferenceCounter, Enumerable.Repeat(item, n)));
+                        break;
+                    }
+                case OpCode.NEWSTRUCT0:
+                    {
+                        Push(new Struct(ReferenceCounter));
+                        break;
+                    }
+                case OpCode.NEWSTRUCT:
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: It's needed `NEWSTRUCT_T`?

File: src/neo-vm/ExecutionEngine.cs [hunk 6]
```diff
@@ -872,26 +931,64 @@ private bool ExecuteInstruction()
...
+                                    if (index < 0) return false;
+                                    Push(index < array.Count);
+                                    break;
+                                }
+                            case Map map:
+                                {
+                                    Push(map.ContainsKey(key));
+                                    break;
+                                }
+                            case Buffer buffer:
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: ByteArray too?

File: src/neo-vm/ExecutionEngine.cs [hunk 7]
```diff
@@ -1060,71 +1108,33 @@ private bool ExecuteInstruction()
...
-                        }
-                        VMArray newArray = new VMArray(ReferenceCounter);
-                        foreach (StackItem item in values)
-                            if (item is Struct s)
-                                newArray.Add(s.Clone());
-                            else
-                                newArray.Add(item);
-                        Push(newArray);
+                        StackItemType type = (StackItemType)instruction.TokenU8;
+                        if (type == StackItemType.Any || !Enum.IsDefined(typeof(StackItemType), type))
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: `type == StackItemType.Any` should be `ISNULL`?

File: src/neo-vm/ExecutionEngine.cs (hunk 4 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: We consider `Buffer` as a reference type, which is `byte[]`, so the default value is `null`.

File: src/neo-vm/ExecutionEngine.cs (hunk 5 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: The types of fields in struct are usually different.

File: src/neo-vm/ExecutionEngine.cs (hunk 6 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: Now we treat `ByteArray` as an immutable `string`.

File: src/neo-vm/ExecutionEngine.cs (hunk 7 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: `Any` is not a real type.

File: src/neo-vm/ExecutionEngine.cs (hunk 6 above)
Comment (shargon) [reply src/neo-vm/ExecutionEngine.cs]: But maybe is interesting to use if we want to know if the length is less than `index`. I think that the use case could be the same as Buffer for `HASKEY`

Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: Done.

**Pull request merged**